From 3fc66965bb552dfc7eae559eb3f69c95e56dfa3d Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Tue, 7 Oct 2025 16:08:09 +0200
Subject: [PATCH 327/553] CMake: Call source_group("Resources") for
 qt_add_resources files

Qt Creator parses the cmake file api response to group a target's
source files into a project tree. CMake has a hardcoded list of file
extensions that it considers as resources, which it places into
a "Resources" source group.
This list is not exhaustive, which causes for example png files to be
under Resources, but not gif files.

Record the list of files added by qt_add_resources into a target
property. When finalizing the target's source groups, place these
files into the "Resources" source group, except for qml and js files
which are kept as source files.

The source_group call respects nested resource file paths, so if a
resources/images/a.png is added, it will be placed under
Resources/resources/images/a.png in the project tree.

Add an opt out variable QT_NO_AUTO_RESOURCE_SOURCE_GROUPS to disable
this behavior if it doesn't fit the project.

Pick-to: 6.8
Fixes: QTBUG-137936
Change-Id: Ic12481be32714cdd95ca33a39e0fc1c802505abd
Reviewed-by: Alexey Edelev <alexey.edelev@qt.io>
(cherry picked from commit 7145c09fc2ef79f167414a44657ed1578e1fa0d7)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/Qt6CoreMacros.cmake | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/corelib/Qt6CoreMacros.cmake b/src/corelib/Qt6CoreMacros.cmake
index 185689623e6..a9c291c66f0 100644
--- a/src/corelib/Qt6CoreMacros.cmake
+++ b/src/corelib/Qt6CoreMacros.cmake
@@ -941,6 +941,11 @@ function(_qt_internal_finalize_source_groups target)
         set(generated_source_group "Source Files/Generated")
     endif()
 
+    get_target_property(resource_source_files "${target}" _qt_resource_source_files)
+    if(NOT resource_source_files)
+        set(resource_source_files "")
+    endif()
+
     foreach(source IN LISTS sources)
         string(GENEX_STRIP "${source}" source)
 
@@ -962,6 +967,9 @@ function(_qt_internal_finalize_source_groups target)
         # due to https://gitlab.kitware.com/cmake/cmake/-/issues/25597
         if(${source_file_path} MATCHES "(\\.qml$)|(\\.js$)")
             source_group("Source Files" FILES ${source_file_path})
+
+            # Remove them from resources files, so they stay as source files.
+            list(REMOVE_ITEM resource_source_files ${source} ${source_file_path})
         endif()
 
         get_source_file_property(is_generated "${source_file_path}" GENERATED)
@@ -969,6 +977,10 @@ function(_qt_internal_finalize_source_groups target)
             source_group(${generated_source_group} FILES ${source_file_path})
         endif()
     endforeach()
+
+    if(NOT QT_NO_AUTO_RESOURCE_SOURCE_GROUPS)
+        source_group("Resources" FILES ${resource_source_files})
+    endif()
 endfunction()
 
 function(_qt_internal_darwin_permission_finalizer target)
@@ -2466,6 +2478,9 @@ function(_qt_internal_process_resource target resourceName)
         _qt_internal_expose_source_file_to_ide(${target} "${file}")
     endforeach()
 
+    set_property(TARGET ${target}
+        APPEND PROPERTY _qt_resource_source_files ${resource_files})
+
     # </qresource></RCC>
     string(APPEND qrcContents "  </qresource>\n</RCC>\n")
 
-- 
2.51.2

